### DISCUSSION TITLE
Bug? Named windows must be globally unique and are globally accessible.

### DISCUSSION THREAD

**Discussion by elyobo** [duckdb/duckdb#5607] [Q&A] [answered]
An error is reported if the same window name is used even in separate contexts, e.g. different CTEs.

This example shows the problem - fails with an error

```sql
CREATE TABLE a (id INT);

WITH
  cte_a AS (
    SELECT *
    FROM a
    WINDOW my_window AS ()
  ),
  cte_b AS (
    SELECT *
    FROM a
    WINDOW my_window AS ()
  )

SELECT *
FROM cte_a CROSS JOIN cte_b;
```

This is, to me, even more strange - it runs! And yet a window named in one CTE might be over a completely different set of data in the other CTE.

```sql
CREATE TABLE a (id INT);

WITH
  cte_a AS (
    SELECT *
    FROM a
    WINDOW my_window AS ()
  ),
  cte_b AS (
    SELECT *, COUNT(*) OVER my_window AS n_rows,
    FROM a
  )

SELECT *
FROM cte_a CROSS JOIN cte_b;
```

One benefit of CTEs is that, with nesting and named windows, they're safely isolated so code can be dynamically composed - window definitions being global, rather than scoped to the subquery/CTE that they're in seems problematic!

**Comment by hannes** [comment, marked as answer]
Indeed those should probably be local to the CTE.

**Reply by elyobo** [reply]
Should I open a bug for this?

**Reply by Mytherin** [reply]
Yes, please do

**Reply by elyobo** [reply]
Done, hope that's helpful - https://github.com/duckdb/duckdb/issues/5609
